rfd/0078-role-variables.md

```diff
@@ -0,0 +1,439 @@
+      values:
+          out: ["foo"]
+    - name: var_b
+      values:
+          out:
+            # var_b can depend on var_a because it is defined above.
+            - "{{vars.var_a}}"
+            # Error: cannot depend on a variable defined later in the list.
+            # - "{{vars.var_c}}"
+    - name: var_c
+      values:
+          # Variables can be used as inputs.
+          input: "{{var_a}}"
+          match: "^foo$"
+          out:
+            - "bar"
+            - "{{vars.var_b}}"
+```
+
+### Trusted clusters
+
+Variables defined in the root cluster will not be usable in leaf clusters.
+It is expected that the variable will be redefined in the leaf cluster.
+
+It's common for deployments to copy the exact same role definitions to all
+clusters, so this seems like the most logical choice.
+As long as the role which defines the variable is present in both clusters and
+it is mapped to its leaf equivalent, you can seamlessly use the variable in both
+clusters.
+
+## Examples
+
+### Defined allow labels for each group
+
+If the Teleport admin cannot create custom traits in their IDP, then it is
+extremely cumbersome and repetitive (borderline impossible) to define a custom
+mapping per group.
+
+Solution today:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  allow:
+    node_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+    app_labels:
+      env:
+        - regexp.replace(external.groups, "dev", "dev")
+        - regexp.replace(external.groups, "dev", "staging")
+        - regexp.replace(external.groups, "qa", "qa")
+        - regexp.replace(external.groups, "qa", "staging")
+```
+
+With more roles and more complex rules, this would be even more difficult. The
+full list would have to be repeated for every value in every role. I don't think
+anyone even knows you can do this, and I hope no-one actually does, because it's
+ridiculous.
+
+With role variables it becomes trivial:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: users
+spec:
+  vars:
+    - name: allow_envs
+      values:
+        - input: "{{external.groups}}"
+          match: "^(qa|devs)$"
+          out:
+            - "$1"
+            - "staging"
+  allow:
+    node_labels:
+      env: "{{vars.allow_envs}}"
+    app_labels:
+      env: "{{vars.allow_envs}}"
+```
+
+The same variable can easily be defined once, and used many times within this
+role.
+
+## Alternatives
+
+### Trait transformations
+
+Instead of defining variables in the role spec, we could create a new `trait`
+resource which can define new traits which will be computed during login and
+embedded in the user's cert.
+
+These traits would be reusable across roles and trusted clusters.
+
+The syntax could be similar to the proposed role variable syntax:
+
+```yaml
```

Comment:
@nklaassen @zmb3 some thoughts ahead of our conversation today:

* Let's pursue traits approach + predicate instead of role variables. This will solve multiple problems for us, including trusted clusters and other use cases. Let's make sure traits rules are evaluated ahead of OIDC mapping. 
* Let's factor out SSO mapping from connector into SSO to roles mapping rules. Once we complete this, we will end up with a complete system for solving a problem of incomplete data coming from identity providers, something we are struggling with now quite often.